rest_server/IOMgrConnection.go

```diff
@@ -0,0 +1,329 @@
+package main
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"unicode/utf8"
+
+	"github.com/gin-gonic/gin"
+	"github.com/golang/protobuf/proto"
+)
+
+type IOMgrConnection struct {
+	net.Conn
+	pool *IOMgrConnPool
+}
+
+func (IOMgrConn IOMgrConnection) cleanupTCPConn() (err error) {
+
+	var IOMgrChunkSize uint64
+	var cpt int16
+
+	for {
+		// Get Current Row Size
+		if _, IOMgrChunkSize, err = IOMgrConn.readVarint(); err != nil {
+			return errors.New("cleanupTCPConn: unable to read chunk size")
+		}
+		if IOMgrChunkSize == 0 {
+			siodbLoggerPool.Output(DEBUG, "cleanupTCPConn: %d chunks dropped.", cpt)
+			return err
+		}
+		siodbLoggerPool.Output(DEBUG, "cleanupTCPConn: Chunk size to drop: %d.", IOMgrChunkSize)
+		buff := make([]byte, IOMgrChunkSize)
+		_, err = io.ReadFull(IOMgrConn.Conn, buff)
+
+		cpt++
+	}
+
+}
+
+func (IOMgrConn IOMgrConnection) streamJSONPayload(c *gin.Context) (err error) {
+
+	var rawData []byte
+
+	//
+	// Read until EOF and write by chunk as defined in HTTPChunkSize
+	//
+
+	// Read raw payload
+	if len(c.Request.Header.Get("Content-Length")) > 0 { // "Transfer-Encoding: chunked" not set
+		// send full size
+		siodbLoggerPool.Output(DEBUG, "Content-Length: %v", c.Request.Header.Get("Content-Length"))
+	} else { // "Transfer-Encoding: chunked" set
+
+	}
+
+	rawData, err = ioutil.ReadAll(c.Request.Body)
+	siodbLoggerPool.Output(DEBUG, "rawData: %v, %v", rawData, err)
+
+	// Write Payload length
+	var buf [binary.MaxVarintLen32]byte
+	encodedLength := binary.PutUvarint(buf[:], uint64(len(rawData)))
+	siodbLoggerPool.Output(DEBUG, "uint64(len(rawData)): %v", uint64(len(rawData)))
+	_, err = IOMgrConn.Conn.Write(buf[:encodedLength])
+	if nil != err {
+		siodbLoggerPool.Output(ERROR, "err: %v", err)
+	}
+
+	// Write raw payload
+	if _, err = IOMgrConn.Write(rawData); err != nil {
+		c.JSON(500, gin.H{"Error:": fmt.Sprintf("Not able to write payload to IOMgr: %v", err)})
+	}
+
+	// Write 0 to indicate EOF
+	encodedLength = binary.PutUvarint(buf[:], uint64(0))
+	_, err = IOMgrConn.Conn.Write(buf[:encodedLength])
+	if nil != err {
+		siodbLoggerPool.Output(ERROR, "err: %v", err)
+	}
+
+	return nil
+
+}
+
+func (IOMgrConn IOMgrConnection) readChunkedJSON(c *gin.Context, statusCode uint32) (err error) {
+
+	// Stream through HTTP > 1.1
+	// https://stackoverflow.com/questions/29486086/how-http2-http1-1-proxy-handle-the-transfer-encoding
+
+	var IOMgrChunkSize uint64
+	var JSONChunk string
+	var JSONChunkBuff string
+
+	for {
+
+		// Get Current IOMgr chunk Size
+		if _, IOMgrChunkSize, err = IOMgrConn.readVarint(); err != nil {
+			c.JSON(500, gin.H{"Error:": fmt.Sprintf("Not able to read chunk size from IOMgr: %v", err)})
+		}
+		siodbLoggerPool.Output(DEBUG, "IOMgrChunkSize: %v", IOMgrChunkSize)
+		if IOMgrChunkSize == 0 {
+			break
+		}
+
+		// Form JSONChunk
+		IOMgrbuff := make([]byte, IOMgrChunkSize)
```

Comment:
what if chunk size you've received will be 3gb? you will be allocating that much? really?
this should be so that you have buffer like 64К (better, regulated with config parameter), you read up to buffer size, then write and so on until full chunk size reached.

Reply:
Yes but actually the `IOMgrChunkSize` is read from the IOMgr. It is the varint that come from the IOMgr to indicate the size of the next chunk. Meaning that if the chunk size is of 3gb, it is because in the parameter file value of `rest_server.chunk_size` was of 3gb.

Reply:
That's wrong premise. It can be 3 gb due to whatever reasons. You can assume that it communicates with iomgr but you actually can't prove. It supposed to, but who knows what is really there?
This one of important principles that must be used in all distributed systems called "defensive programming". This means whatever input you get from an external system, it must be validated and should not cause a harm or resource overuse. In other words, if for some reason protocol fails or broken or corrupted, the program must survive such things. Also, iomgr doesn't use `rest_server.chunk_size`, it is up to REST Server to form chunks of that length and send to HTTP client.
In this case I can see that you take something received from extenal system and without any validation just allocate memory of that size. 
What I would expect here:
1. You read exactly 32-bit number, not 64-bit
2. You validate it. (max allowed chunk size is 2GB, even not 4 which allowed by unsigned 32-bit int, see Siodb code, but likely we should put this into spec)
3. If chunk size is more that some threshold, it must be read in pieces of that max threshold size - to avoid excessive memory usage.
Please fix.